### PR TITLE
Shadow dom rfc

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

File: rfcs/2025-shadow-dom-support.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+- [useOverlay Click Outside in Shadow-DOM context](https://github.com/adobe/react-spectrum/issues/3970)
+
+We have also had a contribution to solve some of the issues. While this is useful, we would like it to feel less hacky and more importantly, we'd like to incorporate the support into our daily lives in as easy as way as possible.
+
+
+## Detailed Design
+
+As mentioned earlier, there are proposed parts to this initiative:
+
+1. Custom React Testing Library Renderer
```
Comment (ritz078) [inline rfcs/2025-shadow-dom-support.md:42]: We do not write special tests for shadow DOM. All our component testing is done using playwright's component testing. We run the same tests both with and without shadow DOM.

File: rfcs/2025-shadow-dom-support.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+## Drawbacks
+
+One concern is on-going support. Our current use cases do not call for Shadow DOM support, so knowledge on the team is currently thin. It will take time for people to ramp up on skills. Not only that, but it isn't part of our weekly testing and we have no plans for it to be at this time, which means we must rely on unit tests as much as possible.
+
+In addition, contributions may occur for a little while until those teams have their needs met. However, it should be assumed that there will be further work to complete the goals as outlined here.
+
+Another concern is that the current approach is, for lack of a better word, hacky. This is because we are accessing and manipulating the Shadow DOM in ways that it wasn't really intended. If we were to rewrite our library today, there are other ways we'd solve these issues which would respect more of the concept of the Shadow DOM and its purpose. What we do here and now may complicate a future where we have different APIs to support this vision of what support would ideally look like.
```
Comment (ritz078) [inline rfcs/2025-shadow-dom-support.md:98]: I think one way to think about this is supporting different root elements instead of shadow DOM. For example components like modal and popover always treat document.body as their root element. Instead react-aria should provide options to treat any other element as a container element.

File: rfcs/2025-shadow-dom-support.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+This is a backwards compatible change, we should just be extending functionality, not breaking any of it.
+
+## Alternatives
+
+Unknown, haven't done research here yet.
+
+## Open Questions
+
+* How to actually define the limitations of our support? See Introduction, it's missing a final sentence with this information.
```
Comment (ritz078) [inline rfcs/2025-shadow-dom-support.md:111]: I think the best way to approach it is to come up with a setup where all new code is compatible and the existing code can be changed when someone needs it. I think in a while, it will be just about using a set of utilities.

We can start with the getOwnerWindow and getOwnerDocument eslint rule in more packages.

File: rfcs/2025-shadow-dom-support.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+## Help Needed
+
+The biggest help we can receive is tests, either in the form of unit tests or in the form of examples of real life applications/setups that we can turn into unit tests. The more tests we have, the less likely we will break anything moving forward after the initial effort is complete.
```
Comment (ritz078) [inline rfcs/2025-shadow-dom-support.md:115]: We want to avoid maintaining a fork so we will try to do our best here. Right now we use patches which are a nightmare during upgrades.

File: rfcs/2025-shadow-dom-support.md (hunk 1 above)
Comment (snowystinger) [reply rfcs/2025-shadow-dom-support.md:42]: I'm not sure what this comment has to do with us? The RFC is for what we need to do to support shadow dom in our library.

I'm also suggesting that we run the same tests both with and without shadow DOM
> we should create a custom React Testing Library renderer for our unit tests which can wrap each test's rendered dom in a shadow root.
> ... without needing to write many specific tests.

There will need to be specific tests as well, but they can be added as edge cases are discovered. For now I'd like to get all of our current tests running both ways.

File: rfcs/2025-shadow-dom-support.md (hunk 2 above)
Comment (snowystinger) [reply rfcs/2025-shadow-dom-support.md:98]: Possibly, what happens in the case of nested shadow doms though? Would we expect people to `addWindowFocusTracking` equivalent for each root and each hook/function that needs it? Would components need to accept a root element to pass down to hooks? context?

This wouldn't work for closed roots. This is why we want to re-think some of the ways parts of our library work. For example, FocusScope moves focus around for you when it's `contain=true`. This can't move into closed shadow roots, which is why the `video` tag breaks our FocusScope at the moment in certain cases. Instead, if we moved to a sentinel-based approach, we could let the browser handle the tab movement.

File: rfcs/2025-shadow-dom-support.md (hunk 3 above)
Comment (snowystinger) [reply rfcs/2025-shadow-dom-support.md:111]: I think this is ambiguous
> all new code is compatible

Does that mean it works in an application that has a single shadow dom at the root? that it can use `video` anywhere? that it can have 3rd party web components anywhere? that it can have shadow roots anywhere?

What if the new code is a new feature in an existing component or hook that has no support? what if that hook depends on lower hooks that don't have support and changing those hooks would also change how other components work.

It's a lofty goal and I think we should aim for support of everything. But it has to be moderated with the cost of implementation and ongoing support.

---
I don't believe this is true
> it will be just about using a set of utilities

as I mentioned earlier, in order to support closed root (as an example), we have to rethink how many components and hooks are implemented.

---
I agree with this
> We can start with the getOwnerWindow and getOwnerDocument eslint rule in more packages.

File: rfcs/2025-shadow-dom-support.md (hunk 4 above)
Comment (snowystinger) [reply rfcs/2025-shadow-dom-support.md:115]: I'm not sure what this has to do with the "help needed" section? Are you just saying it will be hard to create tests and examples?

File: rfcs/2025-shadow-dom-support.md (hunk 1 above)
Comment (ritz078) [reply rfcs/2025-shadow-dom-support.md:42]: > I'm not sure what this comment has to do with us?

Not much. This was just me sharing how we handle this situation.

File: rfcs/2025-shadow-dom-support.md (hunk 4 above)
Comment (ritz078) [reply rfcs/2025-shadow-dom-support.md:115]: Sorry for not being clear. I meant we will like to help here as much as possible so that we do not have to maintain a fork. At the end of the day, we want shadow DOM support to land up here.

File: rfcs/2025-shadow-dom-support.md (hunk 3 above)
Comment (ritz078) [reply rfcs/2025-shadow-dom-support.md:111]: I agree. If the aim here is to add full support for shadow DOM (which will be great), this isn't as simple as adding utilities. When I say full support, I mean everything just works out of the box irrespective of the where a shadow DOM is located. And that will include much more than some utilities and will involve refactoring of certain sections.

**Comment by snowystinger** [comment]
Linking https://github.com/adobe/react-spectrum/discussions/8858 as another surface area to be aware of

**Comment by ritz078** [comment]
@snowystinger Do we have any clarity on how the team is thinking about the future of Shadow DOM support in this repo?

**Comment by snowystinger** [comment]
> @snowystinger Do we have any clarity on how the team is thinking about the future of Shadow DOM support in this repo?

I'm putting together more of a plan this sprint. However, I've started with adding a couple of lint rules to our repo so we can evaluate what's actually missing in terms of more substantial changes. Those are
https://github.com/adobe/react-spectrum/pull/9485
https://github.com/adobe/react-spectrum/pull/9551

Once those are in place, we'll be able to discuss [fix: add shadow dom support for portal provider (S2 v1.0.0, S1 v3.46.0)](https://github.com/adobe/react-spectrum/pull/9369#top) and [Shadow dom event handling](https://github.com/adobe/react-spectrum/pull/8991#top) more clearly, there will be less noise in the PR.
This is the *plan* part and where the bulk of unknown work is.

We are also working on a change to modals that should lead to a change in FocusScope, one of the larger problem areas I'm aware of for ShadowDOM support, specifically due to closed root components such as the native `video` element. That's here: [wip: Use stacking contexts to determine non-inert elements outside modals](https://github.com/adobe/react-spectrum/pull/8796#top)

We are starting with one assumed Shadow root that the entire application lives in. I've attempted to make our test suite run automatically against this case, however, I ran into a number of issues with jsdom, testing library, and user event shadow support, in addition to some assumptions that our tests already make. So creating tests remains one of the biggest hurdles.

Was there anything in particular you wanted to know more about?

**Comment by ritz078** [comment]
@snowystinger Thanks for the clear breakdown, this is really helpful and gives great context on where the real complexity is.

Starting with lint rules to reduce noise feels like a very pragmatic way to surface the Shadow DOM gaps before moving into heavier changes. 

Looking forward to the proposal. Is there a similar goal or plan around iframe support as well?

**Comment by snowystinger** [comment]
> Is there a similar goal or plan around iframe support as well?

Not currently, though there's already two lint rules around window and document, `no-restricted-globals`. We should try to write some similar rules around accessing other globals for things like typeof comparisons, and I don't recall what the other common non-iframe friendly patterns are that we have right now because I haven't thought about it a long while. Focused on ShadowDOM first. But any tests or code paths which are more iframe friendly but also demonstrably the same as they currently exist would be where i would start focusing efforts.

**Comment by ritz078** [comment]
In addition to the already existing lint rules, if you somehow cover `instanceof Element` checks, then you have covered everything. Maybe we can put all those elements in `no-restricted-globals` too.

**Review by yihuiliao — APPROVED** [review]

**Review by devongovett — APPROVED** [review]
